src/aws-cpp-sdk-core/source/monitoring/HttpClientMetrics.cpp

```diff
@@ -18,64 +18,41 @@ namespace Aws
-            {
-                return HttpClientMetricsType::Unknown;
-            }
-            return it->second;
+        HttpClientMetricsType GetHttpClientMetricTypeByName(const Aws::String& name)
+        {
+             auto it = std::find_if(httpClientMetricsNames.begin(), httpClientMetricsNames.end(), [&](const std::pair<HttpClientMetricsType, const char *>& pair) { return name == pair.second; });
+             if (it == httpClientMetricsNames.end())
+             {
+                 return HttpClientMetricsType::Unknown;
+             }
+             return it->first;
         }
 
         Aws::String GetHttpClientMetricNameByType(HttpClientMetricsType type)
         {
-            //TODO: Make static map, Aws::Map cannot be made static with a customer memory manager as of the moment.
-            Aws::Map<int, Aws::String> metricsTypeToName =
-            {
-                std::pair<int, Aws::String>(static_cast<int>(HttpClientMetricsType::DestinationIp), HTTP_CLIENT_METRICS_DESTINATION_IP),
-                std::pair<int, Aws::String>(static_cast<int>(HttpClientMetricsType::AcquireConnectionLatency), HTTP_CLIENT_METRICS_ACQUIRE_CONNECTION_LATENCY),
-                std::pair<int, Aws::String>(static_cast<int>(HttpClientMetricsType::ConnectionReused), HTTP_CLIENT_METRICS_CONNECTION_REUSED),
-                std::pair<int, Aws::String>(static_cast<int>(HttpClientMetricsType::ConnectLatency), HTTP_CLIENT_METRICS_CONNECTION_LATENCY),
-                std::pair<int, Aws::String>(static_cast<int>(HttpClientMetricsType::RequestLatency), HTTP_CLIENT_METRICS_REQUEST_LATENCY),
-                std::pair<int, Aws::String>(static_cast<int>(HttpClientMetricsType::DnsLatency), HTTP_CLIENT_METRICS_DNS_LATENCY),
-                std::pair<int, Aws::String>(static_cast<int>(HttpClientMetricsType::TcpLatency), HTTP_CLIENT_METRICS_TCP_LATENCY),
-                std::pair<int, Aws::String>(static_cast<int>(HttpClientMetricsType::SslLatency), HTTP_CLIENT_METRICS_SSL_LATENCY),
-                std::pair<int, Aws::String>(static_cast<int>(HttpClientMetricsType::Throughput), HTTP_CLIENT_METRICS_THROUGHPUT),
-                std::pair<int, Aws::String>(static_cast<int>(HttpClientMetricsType::DownloadSpeed), HTTP_CLIENT_METRICS_DOWNLOAD_SPEED),
-                std::pair<int, Aws::String>(static_cast<int>(HttpClientMetricsType::UploadSpeed), HTTP_CLIENT_METRICS_UPLOAD_SPEED),
-                std::pair<int, Aws::String>(static_cast<int>(HttpClientMetricsType::Unknown), HTTP_CLIENT_METRICS_UNKNOWN),
-            };
-
-            auto it = metricsTypeToName.find(static_cast<int>(type));
-            if (it == metricsTypeToName.end())
-            {
-                return HTTP_CLIENT_METRICS_UNKNOWN;
-            }
-            return Aws::String(it->second.c_str());
+            assert(httpClientMetricsNames[static_cast<int>(type)].first == type);
+            return Aws::String(httpClientMetricsNames[static_cast<int>(type)].second);
```

Comment:
would prefer you to use the `std::find` lookup i.e.

```cpp
const auto metricName = std::find_if(httpClientMetricsNames.begin(),
  httpClientMetricsNames.end(),
  [&type](const std::pair<HttpClientMetricsType, const char *>& nameEntry) -> bool { return nameEntry.first == type; });
if(metricName == httpClientMetricsNames.end()) {
  return HTTP_CLIENT_METRICS_UNKNOWN;
}
return metricName->second;
```

as this preserves the behavior from before. i recognize that the map logically doesnt have any illegal access, but this makes sure that there is a defined behavior for when none are found when assertions are disabled.

Reply:
Hi, I was unsure whether you wanted to prevent an array out of bound index, or to prevent someone adding some entries in the constant table not in the correct order.

I have added a check for this first case; I can of course put an `std::find` if you want to prevent against the second case.

Reply:
> or to prevent someone adding some entries in the constant table not in the correct order.

yeah my main concern is that someone edits this out of order, creates a binary incompatibility and it shows up in downstream package managers. I think we can change it ourselves to std::find if we find that is the case down the road, and leave the PR as is for now